src/processor/map/map_copy_from.cpp

```diff
@@ -90,100 +78,43 @@ std::unique_ptr<PhysicalOperator> PlanMapper::mapCopyNodeFrom(LogicalOperator* l
         sharedState->fTable, DEFAULT_VECTOR_CAPACITY /* maxMorselSize */, std::move(copyNode));
 }
 
-std::unique_ptr<PhysicalOperator> PlanMapper::createCopyRelColumnsOrLists(
-    std::shared_ptr<CopyRelSharedState> sharedState, LogicalCopyFrom* copyFrom, bool isColumns,
-    std::unique_ptr<PhysicalOperator> copyRelColumns) {
+std::unique_ptr<PhysicalOperator> PlanMapper::createCopyRel(
+    std::shared_ptr<PartitionerSharedState> partitionerSharedState,
+    std::shared_ptr<CopyRelSharedState> sharedState, LogicalCopyFrom* copyFrom,
+    RelDataDirection direction) {
     auto copyFromInfo = copyFrom->getInfo();
     auto outFSchema = copyFrom->getSchema();
-    auto prevOperator = mapOperator(copyFrom->getChild(0).get());
-    Reader* reader;
-    if (prevOperator->getOperatorType() == PhysicalOperatorType::READER) {
-        reader = reinterpret_cast<Reader*>(prevOperator.get());
-    } else {
-        assert(prevOperator->getChild(0)->getOperatorType() == PhysicalOperatorType::READER);
-        reader = reinterpret_cast<Reader*>(prevOperator->getChild(0));
-    }
-
-    auto tableSchema = reinterpret_cast<RelTableSchema*>(copyFromInfo->tableSchema);
-    auto internalIDDataPos =
+    auto tableSchema = dynamic_cast<RelTableSchema*>(copyFromInfo->tableSchema);
+    auto partitioningIdx = direction == FWD ? 0 : 1;
+    auto maxBoundNodeOffset =
```

Comment:
shouldn't this be done during init global state?

Reply:
Left a TODO for now.